File: src/game/quests/steps.ts
import type { GameDatabase, QuestStepRow } from '../data/types'
import { inventoryCount } from '../production/recipes'
import type { PlayerSave } from '../save/types'
import { eligiblePoolEntries } from '../activity/pools'
import { getSkillProgress } from '../activity/xp'
import {
  formatQuestProgressLine,
  objectiveProgressFromStructured,
  parseNotesObjectives,
  parseStructuredObjectives,
  type QuestProgressLine,
} from './objectives'
import {
  asQuestRows,
  getQuest,
  getQuestProgress,
  questAvailableForSave,
  type QuestRow,
} from './quests'
import { hasQuestFlag } from './progress'
import type { StructuredQuestObjectives } from './types'

export type QuestJournalStepState = 'done' | 'current' | 'header'

export interface QuestJournalStep {
  key: string
  label: string
  state: QuestJournalStepState
}

export function asQuestStepRows(db: GameDatabase): QuestStepRow[] {
  return db.QuestSteps ?? []
}

export function getQuestSteps(db: GameDatabase, questId: string): QuestStepRow[] {
  return asQuestStepRows(db)
    .filter((row) => row['Quest ID'] === questId)
    .sort((left, right) => left['Step Order'] - right['Step Order'])
}

export function questUsesSteps(db: GameDatabase, questId: string): boolean {
  return getQuestSteps(db, questId).length > 0
}

/** Every authored step, marked done — used once the quest is finished. */
export function questCompletedJournal(db: GameDatabase, quest: QuestRow): QuestJournalStep[] {
  return getQuestSteps(db, quest['Quest ID']).map((step) => ({
    key: step['Step ID'],
    label: step['Journal Label'],
    state: 'done' as const,
  }))
}

/** Quest-level Notes plus each authored step. */
export function questObjectiveSources(
  db: GameDatabase,
  quest: QuestRow,
): StructuredQuestObjectives[] {
  return [
    parseStructuredObjectives(quest),
    ...getQuestSteps(db, quest['Quest ID']).map((step) =>
      parseNotesObjectives(step.Notes ?? ''),
    ),
  ]
}

/** First Talk step for an NPC may still use the save-wide `talk:npc` flag. */
function talkProgressForStep(
  db: GameDatabase,
  questId: string,
  counters: Record<string, number>,
  talkKey: string,
  stepId: string,
): number {
  const scoped = Number(counters[`${talkKey}:${stepId}`] ?? 0)
  if (scoped >= 1) return scoped
  const npcId = talkKey.slice('talk:'.length)
  const first = getQuestSteps(db, questId).find((step) =>
    parseNotesObjectives(step.Notes ?? '').talkNpcIds.includes(npcId),
  )
  if (first?.['Step ID'] !== stepId) return 0
  return Number(counters[talkKey] ?? 0)
}

function stepProgressLines(
  db: GameDatabase,
  save: PlayerSave,
  questId: string,
  notes: string,
  stepId?: string,
): QuestProgressLine[] {
  const structured = parseNotesObjectives(notes)
  const counters = getQuestProgress(save, questId).counters ?? {}
  const lines = objectiveProgressFromStructured(db, save, structured, counters).progressLines
  if (!stepId) return lines
  return lines.map((line) => {
    if (!line.key.startsWith('talk:')) return line
    return {
      ...line,
      key: `${line.key}:${stepId}`,
      current: talkProgressForStep(db, questId, counters, line.key, stepId),
    }
  })
}

function isAskAroundSkipped(
  db: GameDatabase,
  save: PlayerSave,
  questId: string,
  notes: string,
): boolean {
  const structured = parseNotesObjectives(notes)
  if (structured.optionalTalkNpcIds.length === 0) return false
  if (hasQuestFlag(save, questId, 'choice:bribe') || hasQuestFlag(save, questId, 'choice:combat')) {
    return true
  }
  const quest = getQuest(db, questId)
  if (!quest) return false
  return questAllStepDelivers(db, quest).some(
    (line) => inventoryCount(save, line.targetId) >= line.quantity,
  )
}

function journalProgressLines(
  db: GameDatabase,
  save: PlayerSave,
  questId: string,
  notes: string,
): QuestProgressLine[] {
  const lines = stepProgressLines(db, save, questId, notes)
  const structured = parseNotesObjectives(notes)
  if (structured.optionalTalkNpcIds.length === 0) return lines
  const counters = getQuestProgress(save, questId).counters ?? {}
  const revealed = structured.optionalTalkNpcIds.some(
    (npcId) => Number(counters[`talk:${npcId}`] ?? 0) >= 1,
  )
  if (revealed) return lines
  return lines.filter(
    (line) =>
      !structured.talkNpcIds.some(
        (npcId) => line.key === `talk:${npcId}` || line.key.startsWith(`talk:${npcId}:`),
      ),
  )
}

export function isStepComplete(
  db: GameDatabase,
  save: PlayerSave,
  questId: string,
  notes: string,
  stepId?: string,
): boolean {
  if (isAskAroundSkipped(db, save, questId, notes)) return true
  const lines = stepProgressLines(db, save, questId, notes, stepId)
  if (lines.length === 0) return true
  return lines.every((line) => line.current >= line.required)
}

export function getCurrentStepIndex(
  db: GameDatabase,
  save: PlayerSave,
  quest: QuestRow,
): number {
  const questId = quest['Quest ID']
  const steps = getQuestSteps(db, questId)
  for (let index = 0; index < steps.length; index += 1) {
    const step = steps[index]!
    if (!isStepComplete(db, save, questId, step.Notes ?? '', step['Step ID'])) return index
  }
  return steps.length
}

export function currentStepTalkKey(
  db: GameDatabase,
  save: PlayerSave,
  quest: QuestRow,
  npcId: string,
): string {
  if (!questUsesSteps(db, quest['Quest ID'])) return `talk:${npcId}`
  const steps = getQuestSteps(db, quest['Quest ID'])
  const step = steps[getCurrentStepIndex(db, save, quest)]
  return step ? `talk:${npcId}:${step['Step ID']}` : `talk:${npcId}`
}

export function questStepJournal(
  db: GameDatabase,
  save: PlayerSave,
  quest: QuestRow,
): QuestJournalStep[] {
  const questId = quest['Quest ID']
  const steps = getQuestSteps(db, questId)
  if (steps.length === 0) return []

  const currentIndex = getCurrentStepIndex(db, save, quest)
  if (currentIndex >= steps.length) {
    return questCompletedJournal(db, quest)
  }

  return steps.slice(0, currentIndex + 1).flatMap((step, index) => {
    const done = index < currentIndex
    const head = {
      key: step['Step ID'],
      label: step['Journal Label'],
      state: (done ? 'done' : 'current') as QuestJournalStepState,
    }
    const progressSource = done
      ? stepProgressLines(db, save, questId, step.Notes ?? '', step['Step ID']).filter((line) =>
          line.key.startsWith('action:'),
        )
      : journalProgressLines(db, save, questId, step.Notes ?? '')
    const progress = progressSource.map((line) => ({
      key: line.key,
      label: formatQuestProgressLine(line),
      state: (line.current >= line.required ? 'done' : 'current') as QuestJournalStepState,
    }))
    return [head, ...progress]
  })
}

/** Skill and prior-quest gates for a quest that has not been accepted yet. */
export function questRequirementJournal(
  db: GameDatabase,
  save: PlayerSave,
  quest: QuestRow,
): QuestJournalStep[] {
  const parsed = parseStructuredObjectives(quest)
  const steps: QuestJournalStep[] = []
  for (const requirement of parsed.requiresSkills) {
    const name =
      db.Skills.find((row) => row['Skill ID'] === requirement.skillId)?.['Display Name'] ??
      requirement.skillId
    const level = getSkillProgress(save, requirement.skillId).level
    steps.push({
      key: `skill:${requirement.skillId}`,
      label: `Reach ${name} ${requirement.level} (${level} / ${requirement.level})`,
      state: level >= requirement.level ? 'done' : 'current',
    })
  }
  for (const requiredQuestId of parsed.requiresQuestIds) {
    const name = getQuest(db, requiredQuestId)?.['Display Name'] ?? requiredQuestId
    const done = getQuestProgress(save, requiredQuestId).status === 'completed'
    steps.push({
      key: `quest:${requiredQuestId}`,
      label: `Complete ${name}`,
      state: done ? 'done' : 'current',
    })
  }
  if (steps.length === 0) return steps
  return [{ key: 'header:required', label: 'Required', state: 'header' }, ...steps]
}

/** Objectives for the active step only, or none once every step is done. */
export function questActiveStepObjectives(
  db: GameDatabase,
  save: PlayerSave,
  quest: QuestRow,
): StructuredQuestObjectives | null {
  const steps = getQuestSteps(db, quest['Quest ID'])
  if (steps.length === 0) return null

  const currentIndex = getCurrentStepIndex(db, save, quest)
  if (currentIndex >= steps.length) return null

  return parseNotesObjectives(steps[currentIndex]!.Notes ?? '')
}

export function questAllStepDelivers(
  db: GameDatabase,
  quest: QuestRow,
): StructuredQuestObjectives['delivers'] {
  const steps = getQuestSteps(db, quest['Quest ID'])
  const delivers: StructuredQuestObjectives['delivers'] = []
  for (const step of steps) {
    delivers.push(...parseNotesObjectives(step.Notes ?? '').delivers)
  }
  return delivers
}

export function questAllStepsComplete(
  db: GameDatabase,
  save: PlayerSave,
  quest: QuestRow,
): boolean {
  const steps = getQuestSteps(db, quest['Quest ID'])
  if (steps.length === 0) return false
  return getCurrentStepIndex(db, save, quest) >= steps.length
}

/** Required or optional Talk on the current step (or quest Notes). */
export function questCanTalkToNpc(
  db: GameDatabase,
  save: PlayerSave,
  quest: QuestRow,
  npcId: string,
): boolean {
  const objectives = questUsesSteps(db, quest['Quest ID'])
    ? questActiveStepObjectives(db, save, quest)
    : parseStructuredObjectives(quest)
  if (!objectives) return false
  const talks =
    objectives.talkNpcIds.includes(npcId) || objectives.optionalTalkNpcIds.includes(npcId)
  if (!talks) return false
  return objectives.holds.every((hold) => inventoryCount(save, hold.targetId) >= hold.quantity)
}

/** True when this NPC still has an unfinished Talk step (or Notes talk). */
export function questNpcHasIncompleteTalk(
  db: GameDatabase,
  save: PlayerSave,
  quest: QuestRow,
  npcId: string,
): boolean {
  const questId = quest['Quest ID']
  const steps = getQuestSteps(db, questId)
  if (steps.length === 0) {
    return parseStructuredObjectives(quest).talkNpcIds.includes(npcId)
  }
  return steps.some(
    (step) =>
      parseNotesObjectives(step.Notes ?? '').talkNpcIds.includes(npcId) &&
      !isStepComplete(db, save, questId, step.Notes ?? '', step['Step ID']),
  )
}

/** Locked nodes stay hidden until a current Visit step (or a standing/unlock check) reveals them. */
export function questRevealsLocation(
  db: GameDatabase,
  save: PlayerSave,
  locationId: string,
): boolean {
  for (const quest of asQuestRows(db)) {
    if (getQuestProgress(save, quest['Quest ID']).status !== 'active') continue
    const step = questActiveStepObjectives(db, save, quest)
    if (step?.visitLocationIds.includes(locationId)) return true
  }
  return false
}

export function questTouchesNpcForSave(
  db: GameDatabase,
  save: PlayerSave,
  quest: QuestRow,
  npcId: string,
): boolean {
  if (quest['NPC ID'] === npcId) {
    const status = getQuestProgress(save, quest['Quest ID']).status
    if (status === 'inactive') return questAvailableForSave(db, save, quest)
    return true
  }

  const meta = parseStructuredObjectives(quest)
  const progress = getQuestProgress(save, quest['Quest ID'])
  if (progress.status !== 'active') return false

  if (meta.turnInNpcId === npcId || meta.choiceNpcId === npcId) {
    if (!questUsesSteps(db, quest['Quest ID'])) return true
    if (questActiveStepObjectives(db, save, quest)?.talkNpcIds.includes(npcId)) return true
    return !questNpcHasIncompleteTalk(db, save, quest, npcId)
  }

  return questCanTalkToNpc(db, save, quest, npcId)
}

/** Quest action counts that belong on this activity's dock card. */
export function questActionProgressForActivity(
  db: GameDatabase,
  save: PlayerSave,
  activityId: string,
): QuestProgressLine[] {
  const activity = db.Activities.find((row) => row['Activity ID'] === activityId)
  const poolId = activity?.['Pool ID']
  const actionIds = new Set<string>()
  if (typeof poolId === 'string' && poolId.length > 0) {
    for (const candidate of eligiblePoolEntries(db, poolId)) {
      actionIds.add(candidate.action['Action ID'])
    }
  }
  if (save.currentActivityId === activityId && save.currentActionId) {
    actionIds.add(save.currentActionId)
  }
  if (actionIds.size === 0) return []

  const lines: QuestProgressLine[] = []
  for (const quest of asQuestRows(db)) {
    const questId = quest['Quest ID']
    if (getQuestProgress(save, questId).status !== 'active') continue
    if (questUsesSteps(db, questId)) {
      for (const step of getQuestSteps(db, questId)) {
        for (const line of stepProgressLines(db, save, questId, step.Notes ?? '', step['Step ID'])) {
          const actionId = line.key.startsWith('action:') ? line.key.slice('action:'.length) : null
          if (actionId && actionIds.has(actionId)) lines.push(line)
        }
      }
    } else {
      for (const line of stepProgressLines(db, save, questId, quest.Notes ?? '')) {
        const actionId = line.key.startsWith('action:') ? line.key.slice('action:'.length) : null
        if (actionId && actionIds.has(actionId)) lines.push(line)
      }
    }
  }
  return lines
}

File: packages/ik_rules/lib/src/quests/steps.dart
import 'package:collection/collection.dart';
import 'package:ik_content/ik_content.dart';

import '../js_compat.dart';
import '../production/recipes.dart';
import '../save/generated/save_models.dart';
import 'objectives.dart';
import 'quests.dart';

class QuestJournalStep {
  const QuestJournalStep({required this.key, required this.label, required this.state});

  final String key;
  final String label;
  final String state;

  Map<String, Object?> toJson() => <String, Object?>{'key': key, 'label': label, 'state': state};
}

List<QuestStepRow> getQuestSteps(GameDatabase db, String questId) {
  return db.questSteps
      .where((row) => row.questId == questId)
      .sorted((left, right) => left.stepOrder.compareTo(right.stepOrder))
      .toList();
}

bool questUsesSteps(GameDatabase db, String questId) {
  return getQuestSteps(db, questId).isNotEmpty;
}

List<StructuredQuestObjectives> questObjectiveSources(GameDatabase db, QuestRow quest) {
  return [
    parseStructuredObjectives(quest),
    ...getQuestSteps(
      db,
      jsString(quest['Quest ID']),
    ).map((step) => parseNotesObjectives(step.notes ?? '')),
  ];
}

/// First Talk step for an NPC may still use the save-wide `talk:npc` flag.
num _talkProgressForStep(
  GameDatabase db,
  String questId,
  Map<String, num> counters,
  String talkKey,
  String stepId,
) {
  final scoped = counters['$talkKey:$stepId'] ?? 0;
  if (scoped >= 1) return scoped;
  final npcId = talkKey.substring('talk:'.length);
  final first = getQuestSteps(db, questId).cast<QuestStepRow?>().firstWhere(
    (step) => parseNotesObjectives(step!.notes ?? '').talkNpcIds.contains(npcId),
    orElse: () => null,
  );
  if (first?.stepId != stepId) return 0;
  return counters[talkKey] ?? 0;
}

List<QuestProgressLine> _scopedTalkLines(
  GameDatabase db,
  String questId,
  List<QuestProgressLine> lines,
  Map<String, num> counters,
  String? stepId,
) {
  if (stepId == null) return lines;
  return [
    for (final line in lines)
      if (line.key.startsWith('talk:'))
        QuestProgressLine(
          key: '${line.key}:$stepId',
          label: line.label,
          current: _talkProgressForStep(db, questId, counters, line.key, stepId),
          required: line.required,
        )
      else
        line,
  ];
}

bool _isStepComplete(
  GameDatabase db,
  PlayerSave save,
  String questId,
  String notes, [
  String? stepId,
]) {
  final structured = parseNotesObjectives(notes);
  final counters = getQuestProgress(save, questId).counters ?? const <String, num>{};
  final progress = objectiveProgressFromStructured(db, save, structured, counters);
  final lines = _scopedTalkLines(db, questId, progress.progressLines, counters, stepId);
  if (lines.isEmpty) return true;
  return lines.every((line) => line.current >= line.required);
}

int getCurrentStepIndex(GameDatabase db, PlayerSave save, QuestRow quest) {
  final questId = jsString(quest['Quest ID']);
  final steps = getQuestSteps(db, questId);
  for (var index = 0; index < steps.length; index++) {
    final step = steps[index];
    if (!_isStepComplete(db, save, questId, step.notes ?? '', step.stepId)) return index;
  }
  return steps.length;
}

String currentStepTalkKey(GameDatabase db, PlayerSave save, QuestRow quest, String npcId) {
  final questId = jsString(quest['Quest ID']);
  if (!questUsesSteps(db, questId)) return 'talk:$npcId';
  final steps = getQuestSteps(db, questId);
  final index = getCurrentStepIndex(db, save, quest);
  if (index < 0 || index >= steps.length) return 'talk:$npcId';
  return 'talk:$npcId:${steps[index].stepId}';
}

const String _citadelQuestId = 'QST-0004';
const String _citadelGuideNpcId = 'NPC-0013';

bool citadelGuideHeard(PlayerSave save) {
  return jsNumber(
        getQuestProgress(save, _citadelQuestId).counters?['talk:$_citadelGuideNpcId'] ?? 0,
      ) >=
      1;
}

List<QuestProgressLine> _stepProgressLines(
  GameDatabase db,
  PlayerSave save,
  String questId,
  String notes,
) {
  final structured = parseNotesObjectives(notes);
  final counters = getQuestProgress(save, questId).counters ?? const <String, num>{};
  return objectiveProgressFromStructured(db, save, structured, counters).progressLines;
}

List<QuestJournalStep> questStepJournal(GameDatabase db, PlayerSave save, QuestRow quest) {
  final questId = jsString(quest['Quest ID']);
  final steps = getQuestSteps(db, questId);
  if (steps.isEmpty) return const <QuestJournalStep>[];

  final currentIndex = getCurrentStepIndex(db, save, quest);
  if (questId == _citadelQuestId && citadelGuideHeard(save) && currentIndex >= 1) {
    final heard = steps.first;
    final remaining = steps.skip(1).expand((step) {
      return _stepProgressLines(db, save, questId, step.notes ?? '').map(
        (line) => QuestJournalStep(
          key: line.key,
          label: line.label,
          state: line.current >= line.required ? 'done' : 'current',
        ),
      );
    });
    return [
      QuestJournalStep(key: heard.stepId, label: heard.journalLabel, state: 'done'),
      ...remaining,
    ];
  }

  if (currentIndex >= steps.length) {
    return steps
        .map((step) => QuestJournalStep(key: step.stepId, label: step.journalLabel, state: 'done'))
        .toList();
  }

  return steps.take(currentIndex + 1).toList().asMap().entries.map((entry) {
    final step = entry.value;
    return QuestJournalStep(
      key: step.stepId,
      label: step.journalLabel,
      state: entry.key < currentIndex ? 'done' : 'current',
    );
  }).toList();
}

StructuredQuestObjectives? questActiveStepObjectives(
  GameDatabase db,
  PlayerSave save,
  QuestRow quest,
) {
  final questId = jsString(quest['Quest ID']);
  final steps = getQuestSteps(db, questId);
  if (steps.isEmpty) return null;

  final currentIndex = getCurrentStepIndex(db, save, quest);
  if (currentIndex >= steps.length) return null;

  return parseNotesObjectives(steps[currentIndex].notes ?? '');
}

List<QuestCounterTarget> questAllStepDelivers(GameDatabase db, QuestRow quest) {
  final questId = jsString(quest['Quest ID']);
  final delivers = <QuestCounterTarget>[];
  for (final step in getQuestSteps(db, questId)) {
    delivers.addAll(parseNotesObjectives(step.notes ?? '').delivers);
  }
  return delivers;
}

bool questAllStepsComplete(GameDatabase db, PlayerSave save, QuestRow quest) {
  final questId = jsString(quest['Quest ID']);
  if (!questUsesSteps(db, questId)) return false;
  return getCurrentStepIndex(db, save, quest) >= getQuestSteps(db, questId).length;
}

/// Required or optional Talk on the current step (or quest Notes).
bool questCanTalkToNpc(GameDatabase db, PlayerSave save, QuestRow quest, String npcId) {
  final objectives = questUsesSteps(db, jsString(quest['Quest ID']))
      ? questActiveStepObjectives(db, save, quest)
      : parseStructuredObjectives(quest);
  if (objectives == null) return false;
  final talks =
      objectives.talkNpcIds.contains(npcId) || objectives.optionalTalkNpcIds.contains(npcId);
  if (!talks) return false;
  return objectives.holds.every((hold) => inventoryCount(save, hold.targetId) >= hold.quantity);
}

/// True when this NPC still has an unfinished Talk step (or Notes talk).
bool questNpcHasIncompleteTalk(GameDatabase db, PlayerSave save, QuestRow quest, String npcId) {
  final questId = jsString(quest['Quest ID']);
  final steps = getQuestSteps(db, questId);
  if (steps.isEmpty) {
    return parseStructuredObjectives(quest).talkNpcIds.contains(npcId);
  }
  return steps.any(
    (step) =>
        parseNotesObjectives(step.notes ?? '').talkNpcIds.contains(npcId) &&
        !_isStepComplete(db, save, questId, step.notes ?? '', step.stepId),
  );
}

bool questTouchesNpcForSave(GameDatabase db, PlayerSave save, QuestRow quest, String npcId) {
  if (quest['NPC ID'] == npcId) {
    final status = getQuestProgress(save, jsString(quest['Quest ID'])).status;
    if (status == 'inactive') return questAvailableForSave(db, save, quest);
    return true;
  }

  final meta = parseStructuredObjectives(quest);
  final progress = getQuestProgress(save, jsString(quest['Quest ID']));
  if (progress.status != 'active') return false;

  if (meta.turnInNpcId == npcId || meta.choiceNpcId == npcId) {
    if (!questUsesSteps(db, jsString(quest['Quest ID']))) return true;
    if (questActiveStepObjectives(db, save, quest)?.talkNpcIds.contains(npcId) ?? false) {
      return true;
    }
    return !questNpcHasIncompleteTalk(db, save, quest, npcId);
  }

  return questCanTalkToNpc(db, save, quest, npcId);
}
